Server Plugin UpstreamAuthority cert-manager

JoshVanL opened the pull request:
<!--
1. If this is your first PR, please read our contributor guidelines
https://github.com/spiffe/spiffe/blob/master/CONTRIBUTING.md
https://github.com/spiffe/spire/blob/main/CONTRIBUTING.md

2. Please remember to include a DCO on every commit (`git commit -s`)
https://github.com/apps/dco
-->

**Pull Request check list**

- [ ] Commit conforms to CONTRIBUTING.md?
- [ ] Proper tests/regressions included?
- [ ] Documentation updated?

**Affected functionality**
<!-- Please provide a description of the affected functionality -->

This PR introduces [cert-manager](https://cert-manager.io) as a server upstream authority plugin. This plugin creates [CertificateRequests](https://cert-manager.io/docs/concepts/certificaterequest/) in the target cluster, configured with the referenced issuer. Once signed by the Issuer it returns the signed certificate and CA.

**Description of change**
<!-- Please provide a description of the change -->

Adds cert-manager as a new UA plugin.
Updates k8s dependencies required when importing cert-manager.
Updates test files and kube controllers with new func signatures from updating dependencies.

Let me know if there is anything missing or would like me to split up the PR.

**Which issue this PR fixes**
<!-- optional. `fixes #<issue number>` format will close an issue when this PR is merged -->
fixes #1796

evan2645 commented on pkg/server/plugin/upstreamauthority/certmanager/certmanager.go:
```diff
@@ -0,0 +1,264 @@
+
+		if cmutil.CertificateRequestIsDenied(cr) {
+			cond := cmutil.GetCertificateRequestCondition(cr, cmapi.CertificateRequestConditionDenied)
+			return fmt.Errorf("created CertificateRequest %s/%s has been denied: %s:%s", cr.Namespace, cr.Name, cond.Reason, cond.Message)
```
We should probably log this error with the appropriate fields, and then return a more succinct error message to core. That way we'll get structured logging since core won't extract all of this stuff out from the error it receives.

Similar comment for all the `fmt.Errorf` messages that are templating details into the error message

evan2645 commented on pkg/server/plugin/upstreamauthority/certmanager/certmanager.go:
```diff
@@ -0,0 +1,264 @@
+	}
+
+	// If the configured issuer did not populate the CA on the request we cannot
+	// build the upstream roots. We can only error here.
```
💯 

Is this something that can happen "normally"? Do we need to document anything about it e.g. our requirements on a certain cert-manager configuration?

JoshVanL replied:
It is up to the cert-manager Issuer implementation as to whether they support setting the CA or not, and we can't programmatically determine this before as they are just stringy references.

For cert-manager core Issuers,Vault and CA support setting this field, but ACME for example doesn't.

evan2645 reviewed (commented):
Thank you so much for this contrib @JoshVanL, we sincerely appreciate it! Very excited to gain support for cert-manager :-D

Sorry for the delay in review - I was out most of last week on vacation. I took a pass today and it looks great, had a few small questions here and there but nothing major.

I'm wondering if we should have an integration test for this. How stable is the CertificateRequest API? Is there anything special we need in terms of role bindings, cert-manager config, etc that we need to make all this work?

Finally, it would be great to hear your feedback on the process of authoring a new plugin. We are putting some attention on that currently and hearing a little about your experience would be helpful.

JoshVanL commented on doc/plugin_server_upstreamauthority_cert_manager.md:
```diff
@@ -0,0 +1,40 @@
+
+This plugin will request a signing certificate from cert-manager via a
+[CertificateRequest](https://cert-manager.io/docs/concepts/certificaterequest/)
+resource. Once the referenced issuer has signed the request,
```
Thanks, I have added the rest :)

JoshVanL commented on doc/plugin_server_upstreamauthority_cert_manager.md:
```diff
@@ -0,0 +1,40 @@
+| ------------------------- | ----------------------------------------------------------------- |
+| kube_config_path          | Path to the kubeconfig used to connect to the Kubernetes cluster. |
+| namespace                 | The namespace to create CertificateRequests for signing.          |
+| issuer_name               | The name of the issuer to reference in CertificateRequests.       |
```
Yes, exactly

JoshVanL commented on go.mod:
```diff
@@ -27,60 +25,59 @@ require (
-	k8s.io/kube-aggregator v0.18.2
-	k8s.io/utils v0.0.0-20201110183641-67b214c5f920
-	sigs.k8s.io/controller-runtime v0.6.0
+	k8s.io/api v0.20.2
```
Indeed it is because of importing the cert-manager repo. After some go mod wrangling, I've struggled to reduce the versions. Happy to instead use an [unstructured client](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/unstructured/unstructured.go) to make things easier? This would resolve having to import `jetstack/cert-manager`.

JoshVanL commented on pkg/server/plugin/upstreamauthority/certmanager/certmanager.go:
```diff
@@ -0,0 +1,264 @@
+)
+
+const (
+	pluginName = "cert-manager"
```
Yep, I went back and forth a few times on this!  Happy to change it either way :smile:

JoshVanL commented on pkg/server/plugin/upstreamauthority/certmanager/certmanager.go:
```diff
@@ -0,0 +1,264 @@
+	}
+	// If no issuer_group given, default to cert-manager.io
+	if len(config.IssuerGroup) == 0 {
+		p.log.Warn("configuration has empty issuer_group property, defaulting to 'cert-manager.io'")
```
Makes sense, I will change the default to a debug.

The defaults are the same that are used in cert-manager so made sense to me to port them here from a UX perspective. Happy to remove them completely though, and just error out if they are not defined.

JoshVanL commented:
Thanks for the review @evan2645 :smile: 

> I'm wondering if we should have an integration test for this. How stable is the CertificateRequest API? Is there anything special we need in terms of role bindings, cert-manager config, etc that we need to make all this work?

Sure thing, will work on an integration test next. The CertificateRequest API is v1 now and should be stable from the client point of view. From a role perspective, the client need only "create" permissions for "CertificateRequests" in the defined namespace. The identity of the client to the API server is important though, as this ties into the [approval mechanism](https://cert-manager.io/docs/concepts/certificaterequest/#userinfo). I'll add a note about that in the documentation.


> Finally, it would be great to hear your feedback on the process of authoring a new plugin. We are putting some attention on that currently and hearing a little about your experience would be helpful.

Sure! I mostly found it fairly simple to know which bits I should be adding in, and didn't have much trouble find things. Knowing what goes into `MintX509CAResponse` was a bit tricky, though that is likely due to me not knowing enough about the project yet.

JoshVanL commented:
@evan2645 Had a go at an integration test for cert-manager, let me know what you think :slightly_smiling_face: 

The go dependency changes are tricky.. Feel that using an unstructured client may be the way to go, wdyt?

evan2645 commented on test/integration/suites/upstream-authority-cert-manager/conf/server/spire-server.yaml:
```diff
@@ -0,0 +1,153 @@
+      ca_ttl = "12h"
+      ca_subject {
+        country = ["GB"]
+        organization = ["cert-manager.io"]
```
I might recommend changing this to something SPIRE specific so that it disambiguates the results when we look at leaf issuer and CA (i.e. we can see clearly that leaf was signed by SPIRE, but the root CA is cert-manager). One failure mode here is that we don't take the Upstream path and instead fall back to SPIRE root for some reason, and this case will be harder to distinguish here if we use cert-manager subject for the SPIRE intermediate.

JoshVanL replied:
No problem, I'll change that back to the standard vaules.

evan2645 commented on test/integration/suites/upstream-authority-cert-manager/03-verify-ca:
```diff
@@ -0,0 +1,29 @@
+log-debug "verifying CA..."
+
+./bin/kubectl exec -n spire $(./bin/kubectl get pod -n spire -o name) -- /opt/spire/bin/spire-server x509 mint -spiffeID spiffe://example.org/ns/foo/sa/bar -write .
+leafURIResult=$(./bin/kubectl exec -n spire $(./bin/kubectl get pod -n spire -o name) -- cat svid.pem | openssl x509 --noout -ext subjectAltName | grep URI | sed 's/^ *//g')
+leafIssuerResult=$(./bin/kubectl exec -n spire $(./bin/kubectl get pod -n spire -o name) -- cat svid.pem | openssl x509 --noout -issuer)
+caSubjectResult=$(./bin/kubectl exec -n spire $(./bin/kubectl get pod -n spire -o name) -- cat bundle.pem | openssl x509 --noout --subject)
```
This PR is based on a commit in our main branch that was published prior to our switch to GitHub Actions ... hence, this PR is currently not triggering CI. We require branches be up-to-date prior to merge, so it's not a problem per se, but if you update your branch now you'll get the CI runners going on this PR which may help for debugging this.

JoshVanL replied:
> These flags are single dash i.e. `-noout`, `-subject`
> 
> When I ran it locally, it also complained about `-ext`. I couldn't find an equivalent flag for this that made sense given what you're trying to do, I normally use `-noout -text`

Thanks, I'll move it over to that. CLIs between OS's is always tricky :sweat_smile:, `sed` comes to mind :laughing: 


> This PR is based on a commit in our main branch that was published prior to our switch to GitHub Actions ... hence, this PR is currently not triggering CI. We require branches be up-to-date prior to merge, so it's not a problem per se, but if you update your branch now you'll get the CI runners going on this PR which may help for debugging this.

Cheers! Once changing to the unstructured client it should make rebasing a bit easier.

evan2645 reviewed (commented):
Woot! This is looking great @JoshVanL I think we are almost there! A few more comments from me, couple things I missed on the last pass :( sorry about that.

> Sure! I mostly found it fairly simple to know which bits I should be adding in, and didn't have much trouble find things. Knowing what goes into MintX509CAResponse was a bit tricky, though that is likely due to me not knowing enough about the project yet.

Aah yes ok ... I can see now that you are using our pre-1.0 protos. We have a new set of protos which are better documented, I wonder if they would have helped. You can check them out [here](https://github.com/spiffe/spire-plugin-sdk/blob/main/proto/spire/plugin/server/upstreamauthority/v1/upstreamauthority.proto), curious if that would have made things better for you. @azdagron do we need to do anything special here since we're still in the middle of this plugin subsystem change?

> Had a go at an integration test for cert-manager, let me know what you think 🙂

Awesome, thank you for this!! One of the challenges we've had is maintaining integration code for things we don't know a lot about. I'm sure you all have this problem too, if you have any pointers they'd be very much welcome :-D in the meantime, this integration test goes a long way in helping to mitigate the overhead.

I had some failures trying to run it locally, all related to openssl invocations of course 😆 I think maybe there are a couple typos there. Otherwise, the approach of calling mint and examining the issuer and the bundle sounds good to me.

> The go dependency changes are tricky.. Feel that using an unstructured client may be the way to go, wdyt?

I'll defer to @azdagron on what he wants to do here. I am not sure what other implications unstructured client may bring

azdagron commented:
>  @azdagron do we need to do anything special here since we're still in the middle of this plugin subsystem change?

Naw, we can convert this to v1 trivially after it is merged.

> I'll defer to @azdagron on what he wants to do here. I am not sure what other implications unstructured client may bring

This isn't the first time folks have wanted to update the k8s client deps. We've generally pushed back when there wasn't a requirement. In this case, the unstructured client seems like a workaround, but it's not clear to me how much work. In the end, latching the client code dependency to our minimum supported version only helps compile time compatibility against the minimum version but says nothing about compatibility with the newest versions. A more robust, comprehensive integration test suite that exercised all of the k8s api interactions with at least the minimum and latest versions would give us better assurances.

All this to say that I'm probably ok if we want to bump this now, unless the unstructured client is a fairly trivial drop-in.

JoshVanL commented on pkg/server/plugin/upstreamauthority/certmanager/certmanager.go:
```diff
@@ -0,0 +1,286 @@
+}
+
+// loadConfig parses and defaults incoming configure requests
+func (p *Plugin) loadConfig(req *spi.ConfigureRequest) (*Config, error) {
```
Thanks, that was an oversight

JoshVanL commented on pkg/server/plugin/upstreamauthority/certmanager/certmanager.go:
```diff
@@ -0,0 +1,286 @@
+			},
+			Request: csrBuf.Bytes(),
+			Usages: []cmapi.KeyUsage{
+				cmapi.UsageDigitalSignature,
+				cmapi.UsageKeyEncipherment,
+				cmapi.UsageCertSign,
```
Sure thing, I've also added UsageCRLSign

JoshVanL commented on pkg/server/plugin/upstreamauthority/certmanager/certmanager.go:
```diff
@@ -0,0 +1,286 @@
+	log.Info("waiting for certificaterequest to be signed")
+
+	// Poll the CertificateRequest until it is signed. If not signed after 300
+	// polls, error.
```
There isn't actually.. we should do something here to clean up.

I think what would be best is that we delete successful CertificateRequests in MintX509CA.

We can also annotate created CRs: `cert-manager.spiffe.io/trust-domain: <GlobalConfig.trust-domain>`. We can either delete these CRs optimistically at the start of MintX509CA or at the end as part of the same cleanup. If MintX509CA always fails, it makes sense to me to delete at the start. Wdyt?

JoshVanL replied:
Some prior art to this:

[cert-manager revision-manager](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/certificates/revisionmanager/revisionmanager_controller.go): This deletes CRs based on a max revision history. This requires running a kube controller which wouldn't be appropriate here AFAIK.
[istio-csr optionally deletes the CR after it being successful](https://github.com/cert-manager/istio-csr/blob/65407367362d369e13683bd554b07077ce48a459/pkg/server/server.go#L158)
[cni-lib which optimistically deletes stale CRs during issuance](https://github.com/cert-manager/csi-lib/blob/c1463aa8cfbdc0bade19696bfb2ee9041d116751/manager/manager.go#L321): This is similar to the approach described above^ which I think is the best.

JoshVanL commented:
@evan2645 @azdagron Thanks both for your reviews!

I'll work on adding the cleanup mechanism described above^ and fixing up the integration openssl commands.

I'm also going to give a go at moving the client to unstructured- this shouldn't be a huge change and resolves the mod issues (also doesn't mean this PR changes the kube controllers elsewhere).

> A more robust, comprehensive integration test suite that exercised all of the k8s api interactions with at least the minimum and latest versions would give us better assurances.

The kube compatibility of this integration is effectively tied to the compatibility of the cert-manager installation. Since we use the v1 API which was released in cert-manager v1.0, we have compatibility with [kube back till v1.11](https://cert-manager.io/docs/installation/supported-releases/#upcoming-releases).

azdagron commented:
> I'm also going to give a go at moving the client to unstructured- this shouldn't be a huge change and resolves the mod issues (also doesn't mean this PR changes the kube controllers elsewhere).

Awesome, I appreciate your willingness! Glad to hear it won't be costly.

> The kube compatibility of this integration is effectively tied to the compatibility of the cert-manager installation. Since we use the v1 API which was released in cert-manager v1.0, we have compatibility with kube back till v1.11.

:heart_eyes: that is really good to hear!

My comment about the integration tests were geared towards SPIRE's k8s integrations in general, not necessarily this particular interaction. I've opened up #2350 to track that separate discussion.

JoshVanL commented:
Hi @evan2645 @azdagron! After struggling when testing using the unstructured client and the controller-runtime fake client, I have opted to go with adding the Go struct types in an internal package. This seems to work well. I have also added a note that this can be swapped out in future after we create a [separate repo with just these types](https://github.com/jetstack/cert-manager/issues/3381). This is something I want to pick up for the cert-manger v1.5 release.

I have added in a cleanup function, which is called on defer when minting. It will delete any CertificateRequests which the SPIRE server is responsible for (the same namespace with the label `cert-manager.spiffe.io/trust-domain=<trust domain>`) and are in a terminal state (either failed, denied, or are signed and ready).

I have expanded the unit testing out quite a bit, and now covers the `MintX509CA` func. The integration test has also been expanded a bit to also check that cleanups are occurring.

azdagron reviewed (commented):
I am so sorry it took me so long to get back on this. I think what you have here is fine. If/when that change comes in that moves these types to another repo, even better, but this doesn't sound too bad to maintain. I'm good with the direction this takes.

@evan2645, any other thoughts? are we ready to take this?

evan2645 reviewed (approved):
This is great, thank you @JoshVanL!
